src/compiler/types.ts

```diff
@@ -623,7 +623,7 @@ namespace ts {
         /*@internal*/ autoGenerateKind?: GeneratedIdentifierKind; // Specifies whether to auto-generate the text for an identifier.
         /*@internal*/ autoGenerateId?: number;                    // Ensures unique generated identifiers get unique names, but clones get the same name.
         isInJSDocNamespace?: boolean;                             // if the node is a member in a JSDoc namespace
-        /*@internal*/ typeArguments?: NodeArray<TypeNode>;        // Only defined on synthesized nodes. Though not syntactically valid, used in emitting diagnostics.
+        /*@internal*/ typeArguments?: NodeArray<TypeNode | TypeParameterDeclaration>; // Only defined on synthesized nodes. Though not syntactically valid, used in emitting diagnostics.

```

Comment:
why are TypeParameterDeclarations here suddenly? Is this related to the 'syntactically invalid' comment?

Reply:
...kinda. The `"syntactically invalid"` comment is referencing us making identifiers with text which is not valid identifier text, like `["foo"]` (and is the root cause of having odd quickinfo like `Foo[["bar"]]` for computed properties sometimes). I don't want to fix that as part of this, though, so just left the comment as a reminder that it was fishy, so it'd be fixed later.

Reply:
Actually, I think the Foo[["bar"]] thing is solved in the dynamicNames PR.

Reply:
I wonder if it would have been better to leverage ExpressionWithTypeArguments rather than add this to Identifier.

Reply:
@rbuckton We have this as part of `Identifier` because we reused `EntityName` for the structure which holds this info (which was not allowed to contain arbitrary expressions). With this change, for quickinfo I instead build a real expression tree (instead of an `EntityName`) in order to get the square bracket display for non-identifier names. So I could use `ExpressionWithTypeArguments` instead... but `Identifier` already had the member, and is in use elsewhere for the capture of type arguments for services already. Additionally, the downside of using `ExpressionWithTypeArguments` right now is that it would get auto-parenthesized when it used as the LHS of a `PropertyAccessExpression` (which is both undesirable in this case and would change our output a ton).

Also, there are really only two reasons I didn't fix computed property name to not double-print brackets: 1. To more accurately mimic the old behavior (which we have recorded in baselines) so this is a bit easier to review, and 2. because I'm not 100% sure on how the contents of those square brackets should be parsed (should they be actually parsed to generate an AST as an isolated expression if not a simple number or string literal? Where do we generate symbol names like this, and what constraints are on them?).